File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
  let maxArea = 0;
  const stack = [];

  for (let i = 0; i < heights.length + 1; i++) {
    const currHeight = i === heights.length ? 0 : heights[i];

    while (stack.length !== 0 && currHeight < heights[stack[stack.length - 1]]) {
      const top = stack.pop();

      const height = heights[top];
      const width = stack.length === 0 ? i : i - stack[stack.length - 1] - 1;

      const currArea = height * width;
      maxArea = Math.max(currArea, maxArea);
    }

    stack.push(i);
  }

  return maxArea;
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        left_bounds = [0] * len(heights)
        right_bounds = [len(heights) - 1] * len(heights)

        for i in range(1, len(heights)):
            curr_height = heights[i]
            search = i
            while search > 0 and heights[search - 1] >= curr_height:
                search = left_bounds[search - 1]
            left_bounds[i] = search

        for i in range(len(heights) - 2, -1, -1):
            curr_height = heights[i]
            search = i
            while search < len(heights) - 1 and heights[search + 1] >= curr_height:
                search = right_bounds[search + 1]
            right_bounds[i] = search

        # print("left_bounds=", left_bounds)
        # print("right_bounds=", right_bounds)

        max_area = 0
        for i in range(len(heights)):
            max_area = max(
                max_area,
                (right_bounds[i] - left_bounds[i] + 1) * heights[i]
            )

        return max_area